source/server/connection_handler_impl.cc

```diff
@@ -69,71 +58,138 @@ void ConnectionHandlerImpl::ActiveListener::removeConnection(ActiveConnection& c
   parent_.num_connections_--;
 }
 
-ConnectionHandlerImpl::ActiveListener::ActiveListener(
-    ConnectionHandlerImpl& parent, Network::ListenSocket& socket,
-    Network::FilterChainFactory& factory, Stats::Scope& scope, uint64_t listener_tag,
-    const Network::ListenerOptions& listener_options)
-    : ActiveListener(
-          parent, parent.dispatcher_.createListener(parent, socket, *this, scope, listener_options),
-          factory, scope, listener_tag) {}
+ConnectionHandlerImpl::ActiveListener::ActiveListener(ConnectionHandlerImpl& parent,
+                                                      Network::ListenerConfig& config)
+    : ActiveListener(parent,
+                     parent.dispatcher_.createListener(config.socket(), *this, config.bindToPort()),
+                     config) {}
 
 ConnectionHandlerImpl::ActiveListener::ActiveListener(ConnectionHandlerImpl& parent,
                                                       Network::ListenerPtr&& listener,
-                                                      Network::FilterChainFactory& factory,
-                                                      Stats::Scope& scope, uint64_t listener_tag)
-    : parent_(parent), factory_(factory), listener_(std::move(listener)),
-      stats_(generateStats(scope)), listener_tag_(listener_tag) {}
+                                                      Network::ListenerConfig& config)
+    : parent_(parent), listener_(std::move(listener)),
+      stats_(generateStats(config.listenerScope())), listener_tag_(config.listenerTag()),
+      config_(config) {}
 
 ConnectionHandlerImpl::ActiveListener::~ActiveListener() {
+  // Purge sockets that have not progressed to connections. This should only happen when
+  // a listener filter stops iteration and never resumes.
+  while (!sockets_.empty()) {
+    ActiveSocketPtr removed = sockets_.front()->removeFromList(sockets_);
```

Comment:
Do you have test coverage for this case?

Reply:
proxy_protocol_test.cc should exercise this, it has tests for invalid/truncated proxy protocol messages.